### PR TITLE
search: add more traces

### PR DESCRIPTION
**Pull request by stefanhengl**
Currently, we only trace file/path searches. This PR adds traces for all results types, at least on the highest level. We also increase the granularity of traces for getting user settings.

Learnings from this change:
* we access user settings twice, before and after search (fix costs, marked in orange)
* repository search (default result-type, marked in red) might be a bottleneck for larger index sizes because repo resolution is on its critical path. Streaming will probably mask this problem though.

There is still a gap after the end of `doResults`. I will look at this next.

image

### CODE REVIEW

**Comment by stefanhengl** [comment]
edit: I also added a trace for dynamic filters which reveals that dynamic filters are on the critical path for search and are responsible for checking user settings after each search request 🤔

File: cmd/frontend/graphqlbackend/search_results.go [hunk 1]
```diff
@@ -224,6 +226,12 @@ var commonFileFilters = []struct {
 }
 
 func (sr *SearchResultsResolver) DynamicFilters(ctx context.Context) []*searchFilterResolver {
+	var err error
+	tr, ctx := trace.New(ctx, "DynamicFilters", "", trace.Tag{Key: "resolver", Value: "SearchResultsResolver"})
+	defer func() {
+		tr.SetError(err)
```
Comment (keegancsmith) [inline cmd/frontend/graphqlbackend/search_results.go:232]: there is no error to set? looks like we never set err. You have done this a few times in this PR. If there isn't an error return in a function I don't think you need seterror (it's a noop if err is nil)

File: cmd/frontend/graphqlbackend/search_results.go [hunk 2]
```diff
@@ -1710,13 +1748,13 @@ func (r *searchResolver) isGlobalSearch() bool {
...
-func (r *searchResolver) doResults(ctx context.Context, forceOnlyResultType string) (*SearchResultsResolver, error) {
-	var err error
-	tr, ctx := trace.New(ctx, "graphql.SearchResults", r.rawQuery())
+func (r *searchResolver) doResults(ctx context.Context, forceOnlyResultType string) (_ *SearchResultsResolver, err error) {
+	tr, ctx := trace.New(ctx, "doResults", "")
 	defer func() {
 		tr.SetError(err)
 		tr.Finish()
 	}()
+	tr.LogFields(otlog.String("query", r.rawQuery()))
```
Comment (keegancsmith) [inline cmd/frontend/graphqlbackend/search_results.go:1757]: why did you move the query out of the title?

**Review by keegancsmith — APPROVED** [review]

File: cmd/frontend/graphqlbackend/search_results.go (hunk 2 above)
Comment (stefanhengl) [reply cmd/frontend/graphqlbackend/search_results.go:1757]: it feels odd to have it as "title", which is nothing else than a tag anyway. I am basically just renaming the tag from "title" to "query". Maybe there are side effects with net trace that I don't know?

File: cmd/frontend/graphqlbackend/search_results.go (hunk 1 above)
Comment (stefanhengl) [reply cmd/frontend/graphqlbackend/search_results.go:232]: All of the functions have some kind of internal error that they don't return or merge into `multiErr`. I wanted to add this information to the trace.

Comment (stefanhengl) [reply cmd/frontend/graphqlbackend/search_results.go:232]: Edit: on second thought, you are right. The functions that return the error should tag their spans as an error, not the calling function.

File: cmd/frontend/graphqlbackend/search_results.go (hunk 2 above)
Comment (keegancsmith) [reply cmd/frontend/graphqlbackend/search_results.go:1757]: yes it affects net/trace. the summary display shows it, very useful to see the trace there. But jaeger is pretty well itnegrated these days, so net/trace is less important than it used to be.

Comment (stefanhengl) [reply cmd/frontend/graphqlbackend/search_results.go]: understand, I will put the query back where it was. I should have a look at net/trace before I make further changes :-)
